// @builder.io/eslint-plugin-mitosis/dist/rules/__tests__/jsx-callback-arg-name.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
exports.__esModule = true;
var eslint_1 = require("eslint");
var jsx_callback_arg_name_1 = require("../jsx-callback-arg-name");
var opts = {
    filename: 'component.lite.tsx',
    parserOptions: {
        ecmaVersion: 2018,
        sourceType: 'module',
        ecmaFeatures: {
            jsx: true
        }
    }
};
var ruleTester = new eslint_1.RuleTester();
ruleTester.run('jsx-callback-arg-name', jsx_callback_arg_name_1["default"], {
    valid: [
        __assign(__assign(__assign({}, opts), { code: '<button/>' }), opts),
        __assign(__assign(__assign({}, opts), { code: '<button type="button"/>' }), opts),
        __assign(__assign(__assign({}, opts), { code: '<button onClick={ null }/>' }), opts),
        __assign(__assign(__assign({}, opts), { code: '<button onClick={ "string" }/>' }), opts),
        __assign(__assign({}, opts), { code: '<button onClick={ event => doSomething(event) }/>' }),
        __assign(__assign(__assign({}, opts), { code: '<button onClick={ () => doSomething() }/>' }), opts),
        __assign(__assign(__assign({}, opts), { code: '<button onClick={ function(event) {} }/>' }), opts),
        __assign(__assign({}, opts), { code: '<button onClick={ e => console.log(e) }/>', filename: 'file.jsx' }),
    ],
    invalid: [
        __assign(__assign({}, opts), { code: '<button onClick={ e => console.log(e) }/>', output: '<button onClick={ event => console.log(e) }/>', errors: ['Callback parameter must be called `event`'] }),
        __assign(__assign({}, opts), { code: '<button onClick={ function (e) {} }/>', output: '<button onClick={ function (event) {} }/>', errors: ['Callback parameter must be called `event`'] }),
        __assign(__assign({}, opts), { code: '<button onClick={ function foobar(e) {} }/>', output: '<button onClick={ function foobar(event) {} }/>', errors: ['Callback parameter must be called `event`'] }),
    ]
});
