// @builder.io/eslint-plugin-mitosis/dist/rules/__tests__/no-assign-props-to-state.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
exports.__esModule = true;
var eslint_1 = require("eslint");
var no_assign_props_to_state_1 = require("../no-assign-props-to-state");
var opts = {
    filename: 'component.lite.tsx',
    parserOptions: {
        ecmaVersion: 2018,
        sourceType: 'module',
        ecmaFeatures: {
            jsx: true
        }
    }
};
var ruleTester = new eslint_1.RuleTester();
ruleTester.run('no-assign-props-to-state', no_assign_props_to_state_1["default"], {
    valid: [
        __assign(__assign({}, opts), { code: "\n      import { useStore } from '@builder.io/mitosis';\n      export default function MyComponent(props) {\n        const state = useStore({ text: null });\n\n        onMount(() => {\n          state.text = props.text;\n        });\n      }\n      " }),
        __assign(__assign({}, opts), { code: "\n      import { useStore } from '@builder.io/mitosis';\n      import { foo } from '../helpers';\n\n      export default function MyComponent(props) {\n        const state = useStore({ \n          text: null,\n          fn1() {\n            return foo(props.text);\n          },\n          fn2() {\n            return foo({ text: props.text });\n          }\n        });\n\n        onMount(() => {\n          state.text = props.text;\n        });\n      }\n      " }),
        __assign(__assign({}, opts), { code: "\n      import { useStore } from '@builder.io/mitosis';\n\n      export default function MyComponent(props) {\n        const state = useStore({ text: props.text });\n      }\n      ", filename: 'file.jsx' }),
    ],
    invalid: [
        __assign(__assign({}, opts), { code: "\n      import { useStore } from '@builder.io/mitosis';\n      export default function MyComponent(props) {\n        const state = useStore({ text: props.text });\n      }\n      ", errors: ['"props" can\'t be assign to  to "state" directly'] }),
    ]
});
