// @builder.io/mitosis-cli/dist/build/helpers/transpile.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transpileIfNecessary = exports.transpile = exports.transformImports = void 0;
var mitosis_1 = require("@builder.io/mitosis");
var esbuild = __importStar(require("esbuild"));
/**
 * Remove `.lite` or `.svelte` extensions from imports without having to load a slow parser like babel
 * E.g.
 *
 * convert `import { foo } from './block.lite';` -> `import { foo } from './block';`
 *
 * convert `import { foo } from './block.svelte';` -> `import { foo } from './block';`
 */
var transformImports = function (_a) {
    var target = _a.target, options = _a.options;
    return function (code) {
        var _a, _b, _c, _d;
        // we start by replacing all `context.lite` imports with `context`
        // This Context replace is only needed for non-mitosis components, i.e. plain `.js`/`.ts` files.
        // Mitosis components have logic that transform context import paths correctly.
        code = code.replace(/\.context\.lite(.js|.ts)?(['"])/g, ".context.js$2");
        // afterwards, we replace all component imports with the correct file extension
        code = (0, mitosis_1.renameComponentImport)({
            importPath: code,
            target: target,
            explicitImportFileExtension: ((_b = (_a = options.options) === null || _a === void 0 ? void 0 : _a[target]) === null || _b === void 0 ? void 0 : _b.explicitImportFileExtension) || false,
        });
        // if we really need to update the file extensions as well from .js to something else we do it here
        code = (0, mitosis_1.renameImport)({
            importPath: code,
            target: target,
            explicitImportFileExtension: ((_d = (_c = options.options) === null || _c === void 0 ? void 0 : _c[target]) === null || _d === void 0 ? void 0 : _d.explicitImportFileExtension) || false,
        });
        return code;
    };
};
exports.transformImports = transformImports;
/**
 * Runs `esbuild` on a file, and performs some additional transformations.
 */
var transpile = function (_a) {
    var path = _a.path, content = _a.content, target = _a.target, options = _a.options;
    return __awaiter(void 0, void 0, void 0, function () {
        var transpilerOptions, format, output, e_1;
        var _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _c.trys.push([0, 2, , 3]);
                    transpilerOptions = (_b = options.options[target]) === null || _b === void 0 ? void 0 : _b.transpiler;
                    format = (transpilerOptions === null || transpilerOptions === void 0 ? void 0 : transpilerOptions.format) || 'esm';
                    return [4 /*yield*/, esbuild.transform(content, {
                            format: format,
                            /**
                             * Collisions occur between TSX and TS Generic syntax. We want to only provide this loader config if the file is
                             * a mitosis `.lite.tsx` file.
                             */
                            loader: (0, mitosis_1.checkIsLiteComponentFilePath)(path) ? 'tsx' : 'ts',
                            target: 'es6',
                        })];
                case 1:
                    output = _c.sent();
                    if (output.warnings.length) {
                        console.warn("Warnings found in file: ".concat(path), output.warnings);
                    }
                    return [2 /*return*/, output.code];
                case 2:
                    e_1 = _c.sent();
                    console.error("Error found in file: ".concat(path));
                    throw e_1;
                case 3: return [2 /*return*/];
            }
        });
    });
};
exports.transpile = transpile;
var transpileIfNecessary = function (_a) {
    var content = _a.content, options = _a.options, path = _a.path, target = _a.target;
    return __awaiter(void 0, void 0, void 0, function () {
        var _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    if (!(0, mitosis_1.checkShouldOutputTypeScript)({ target: target, options: options })) return [3 /*break*/, 1];
                    _b = content;
                    return [3 /*break*/, 3];
                case 1: return [4 /*yield*/, (0, exports.transpile)({ path: path, target: target, options: options, content: content })];
                case 2:
                    _b = _c.sent();
                    _c.label = 3;
                case 3: return [2 /*return*/, _b];
            }
        });
    });
};
exports.transpileIfNecessary = transpileIfNecessary;
