// @builder.io/eslint-plugin-mitosis/dist/rules/no-state-destructuring.js
"use strict";
exports.__esModule = true;
var types = require("@babel/types");
var isMitosisPath_1 = require("../helpers/isMitosisPath");
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
var rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'disallow destructuring state',
            recommended: true
        }
    },
    create: function (context) {
        // variables should be defined here
        var filename = context.getFilename();
        if (!(0, isMitosisPath_1["default"])(filename))
            return {};
        // ----------------------------------------------------------------------
        // Helpers
        // ----------------------------------------------------------------------
        // any helper functions should go here or else delete this section
        // ----------------------------------------------------------------------
        // Public
        // ----------------------------------------------------------------------
        //
        var listener = {
            VariableDeclarator: function (node) {
                if (!types.isObjectPattern(node.id) ||
                    !types.isIdentifier(node.init) ||
                    node.init.name !== 'state')
                    return;
                context.report({
                    node: node,
                    message: "destructuring state isn't allowed: use standard assignment instead"
                });
            }
        };
        return listener;
    }
};
exports["default"] = rule;
