// @builder.io/eslint-plugin-mitosis/dist/rules/no-var-declaration-or-assignment-in-component.js
"use strict";
exports.__esModule = true;
var ts_pattern_1 = require("ts-pattern");
var hooks_1 = require("../constants/hooks");
var isMitosisPath_1 = require("../helpers/isMitosisPath");
var noOp_1 = require("../helpers/noOp");
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
var rule = {
    meta: {
        type: 'suggestion',
        docs: {
            description: 'disallow variable declaration inside component file.',
            recommended: true
        }
    },
    create: function (context) {
        // variables should be defined here
        var filename = context.getFilename();
        if (!(0, isMitosisPath_1["default"])(filename))
            return {};
        // ----------------------------------------------------------------------
        // Helpers
        // ----------------------------------------------------------------------
        // any helper functions should go here or else delete this section
        // ----------------------------------------------------------------------
        // Public
        // ----------------------------------------------------------------------
        //
        var listener = {
            VariableDeclaration: function (node) {
                (0, ts_pattern_1.match)(node)["with"]({
                    declarations: [
                        (0, ts_pattern_1.not)({
                            init: {
                                type: 'CallExpression',
                                callee: {
                                    type: 'Identifier',
                                    name: (0, ts_pattern_1.when)(function (v) {
                                        return v === hooks_1.HOOKS.STATE ||
                                            v === hooks_1.HOOKS.CONTEXT ||
                                            v === hooks_1.HOOKS.STORE ||
                                            v === hooks_1.HOOKS.REF;
                                    })
                                }
                            }
                        }),
                    ],
                    parent: {
                        type: 'BlockStatement',
                        parent: {
                            type: 'FunctionDeclaration',
                            parent: {
                                type: 'ExportDefaultDeclaration'
                            }
                        }
                    }
                }, function (node) {
                    context.report({
                        node: node,
                        message: 'Variable declaration inside component is ignored during compilation'
                    });
                })
                    .otherwise(noOp_1["default"]);
            },
            AssignmentExpression: function (node) {
                (0, ts_pattern_1.match)(node)["with"]({
                    parent: {
                        type: 'ExpressionStatement',
                        parent: {
                            type: 'BlockStatement',
                            parent: {
                                type: 'FunctionDeclaration',
                                parent: {
                                    type: 'ExportDefaultDeclaration'
                                }
                            }
                        }
                    }
                }, function (node) {
                    context.report({
                        node: node,
                        message: 'Variable assignment inside component is ignored during compilation'
                    });
                })
                    .otherwise(noOp_1["default"]);
            }
        };
        return listener;
    }
};
exports["default"] = rule;
