// @builder.io/eslint-plugin-mitosis/dist/rules/no-var-declaration-in-jsx.js
"use strict";
exports.__esModule = true;
var types = require("@babel/types");
var isMitosisPath_1 = require("../helpers/isMitosisPath");
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
var rule = {
    meta: {
        type: 'suggestion',
        docs: {
            description: 'disallow variable declarations inside jsx.',
            recommended: true
        }
    },
    create: function (context) {
        // variables should be defined here
        var filename = context.getFilename();
        if (!(0, isMitosisPath_1["default"])(filename))
            return {};
        // ----------------------------------------------------------------------
        // Helpers
        // ----------------------------------------------------------------------
        // any helper functions should go here or else delete this section
        // ----------------------------------------------------------------------
        // Public
        // ----------------------------------------------------------------------
        //
        var listener = {
            VariableDeclarator: function (node) {
                var ans = context.getAncestors();
                if (ans.find(types.isJSXElement) && !ans.find(types.isJSXAttribute)) {
                    context.report({
                        node: node,
                        message: 'Variable declaration inside jsx is ignored during compilation'
                    });
                }
            }
        };
        return listener;
    }
};
exports["default"] = rule;
