// @builder.io/mitosis-cli/dist/commands/new.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var command = {
    name: 'new',
    alias: 'n',
    description: 'mitosis new [options]',
    run: function (toolbox) {
        return __awaiter(this, void 0, void 0, function () {
            function exec(cmd, opts) {
                return __awaiter(this, void 0, void 0, function () {
                    var result, e_1;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                _a.trys.push([0, 2, , 3]);
                                return [4 /*yield*/, sys.exec(cmd, opts)];
                            case 1:
                                result = _a.sent();
                                result.stdout && print.info(result.stdout);
                                return [3 /*break*/, 3];
                            case 2:
                                e_1 = _a.sent();
                                print.error("Command failed with exit code ".concat(e_1.exitCode, ": ").concat(e_1.command));
                                e_1.stdout && print.error(e_1.stdout);
                                e_1.stderr && print.error(e_1.stderr);
                                process.exit(1);
                                return [3 /*break*/, 3];
                            case 3: return [2 /*return*/];
                        }
                    });
                });
            }
            var sys, pkg, print, spinner;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        sys = toolbox.system;
                        pkg = toolbox.packageManager;
                        print = toolbox.print;
                        spinner = print.spin({});
                        spinner.start('Creating new project');
                        return [4 /*yield*/, exec('npm init -y')];
                    case 1:
                        _a.sent();
                        spinner.succeed('Wrote package.json');
                        spinner.start('Installing packages');
                        return [4 /*yield*/, pkg.add(['@builder.io/mitosis', '@builder.io/mitosis-cli', 'typescript'], {
                                dev: true,
                                force: 'npm',
                            })];
                    case 2:
                        _a.sent();
                        spinner.succeed('Installed packages');
                        toolbox.template.generate({
                            template: 'tsconfig.json.ejs',
                            target: 'tsconfig.json',
                        });
                        spinner.succeed('Wrote tsconfig.json');
                        toolbox.template.generate({
                            template: 'mitosis.config.js.ejs',
                            target: 'mitosis.config.js',
                        });
                        spinner.succeed('Wrote mitosis.config.js ');
                        toolbox.template.generate({
                            template: 'component.lite.tsx.ejs',
                            target: 'src/component.lite.tsx',
                        });
                        spinner.succeed('Wrote src/component.lite.tsx');
                        spinner.stopAndPersist({ text: 'Done!' });
                        return [2 /*return*/];
                }
            });
        });
    },
};
exports.default = command;
